add optional bones messages for player and log

poikilos opened the pull request:
I have been using this on my server and thought other people might want it (feedback is welcome), especially since it is optional.
The changes include two options (for minetest.conf) which can be enabled separately, or both can be enabled:
* `enable_bones_logging` (boolean, default false): When a player dies, the code logs a message regarding whether bones were dropped (or why they werent), and the location. The bones logging option could be helpful for similar reasons as existing log messages generated by minetest (forensics, debugging, etc). You may want to consider making this option on by default (true if the minetest.conf setting does not exist).
* `enable_bones_chat_msg` (boolean, default false): When a player dies, that player gets a chat message with the same information as above. The chat option has helped reduce support requests for me, but I can understand some server owners may not want it if they believe it detracts too much from immersion or from players learning minetest navigation skills. In my case, I want to encourage building (and reduce the fallout of gang/repeat killing), so if a player dies, I'd rather have them be able to find where they left off. Even just informing them of why their bones didn't drop helps reduce support requests since otherwise it can be perceived as a bug (or as though they failed to find them and should keep looking for something that doesn't exist).

poikilos commented:
Can someone explain why travis-ci complains? The code I added is very straightforward and has been in use.

rubenwardy commented:
Some of your lines are too long, as you're shadowing a variable - `pos`.

See the travis logs by clicking the x

SmallJoker commented on mods/bones/init.lua:
```diff
@@ -186,18 +186,39 @@ minetest.register_on_dieplayer(function(player)
 		bones_mode = "bones"
 	end
 
+	local enable_bones_logging = minetest.settings:get_bool("enable_bones_logging")
```
`minetest.settings:get_bool("enable_bones_logging") ~= false` if you want to default this setting to `true`

SmallJoker replied:
Little addition: I commented on this line because you're specifying the `enable_bones_logging` default value to `true in the settingtypes.txt file but that's (yet) only used for the Advanced Settings GUI in the main menu, so it doesn't have an effect on the actual defaults.

SmallJoker commented on mods/bones/init.lua:
```diff
@@ -186,18 +186,39 @@ minetest.register_on_dieplayer(function(player)
 
+	local enable_bones_logging = minetest.settings:get_bool("enable_bones_logging")
+	local enable_bones_chat_msg = minetest.settings:get_bool("enable_bones_chat_msg")
+	local pos = vector.round(player:getpos())
```
`player:get_pos()`, the other function name is deprecated.

SmallJoker commented on mods/bones/init.lua:
```diff
@@ -186,18 +186,39 @@ minetest.register_on_dieplayer(function(player)
+		if enable_bones_logging then
+			minetest.log("action", "[bones] " .. player:get_player_name() ..
+				"'s bones do not remain since player has creative -- died at " ..
+				minetest.pos_to_string(vector.round(player:getpos())))
```
You may use `pos` here for shorter and faster code

SmallJoker commented on mods/bones/init.lua:
```diff
@@ -186,18 +186,39 @@ minetest.register_on_dieplayer(function(player)
+		end
+		if enable_bones_chat_msg then
+			minetest.chat_send_player(player:get_player_name(), player:get_player_name() ..
+				"'s bones do not remain since player has creative -- died at " .. minetest.pos_to_string(pos))
```
`pos` here too. Line is too long (more than 80 characters), please break it again.

SmallJoker commented on mods/bones/init.lua:
```diff
@@ -186,18 +186,39 @@ minetest.register_on_dieplayer(function(player)
+		if enable_bones_logging then
+			minetest.log("action", "[bones] " .. player:get_player_name() ..
+				"'s bones do not remain since player has no items -- died at " ..
+				minetest.pos_to_string(vector.round(player:getpos())))
```
`pos`

SmallJoker commented on mods/bones/init.lua:
```diff
@@ -186,18 +186,39 @@ minetest.register_on_dieplayer(function(player)
+		end
+		if enable_bones_chat_msg then
+			minetest.chat_send_player(player:get_player_name(), player:get_player_name() ..
+			"'s bones do not remain since player has no items -- died at " .. minetest.pos_to_string(pos))
```
Missing indent and line too long again.

poikilos commented:
@SmallJoker the purpose of logging "bones do not remain since player has no items" is to also log "died at". Though from a purely data perspective only position (and hp) has changed, from perspective of gameplay a die event is surely as significant as a dig event, I would suggest.

SmallJoker commented:
Thanks for your explanation. Logging all deaths is also consistent, I see.
Our code style guidelines can be found on the development wiki: http://dev.minetest.net/Lua_code_style_guidelines (same as the [C(++) code style](http://dev.minetest.net/Code_style_guidelines))

SmallJoker commented on settingtypes.txt:
```diff
@@ -46,3 +46,12 @@ share_bones_time (Bone share time) int 1200 0
 enable_stairs_replace_abm (Replace old stairs) bool false
+
+
+#    If enabled, the reason why bones where dropped or not dropped, along
```
I suggest rephrasing: `Logs the reason and location of newly placed bones as action.`
otherwise, `where` should be `were` to make sense in this context.

SmallJoker commented on settingtypes.txt:
```diff
@@ -46,3 +46,12 @@ share_bones_time (Bone share time) int 1200 0
+#    with death location, is written to the minetest log.
+enable_bones_logging (Log location of deaths and reason if no bones) bool true
+
+#    If enabled, the reason why bones where dropped or not dropped, along
```
Possible rephrasing here too: `Sends the player a message about the condition and location of their bones on death.`

SmallJoker commented on settingtypes.txt:
```diff
@@ -46,3 +46,12 @@ share_bones_time (Bone share time) int 1200 0
+
+#    If enabled, the reason why bones where dropped or not dropped, along
+#    with death location, is sent as a chat message to player.
+enable_bones_chat_msg (Send player last death location and reason if no bones) bool false
```
These short descriptions should be shorter. How about: `Inform player about bones`?

SmallJoker commented on settingtypes.txt:
```diff
@@ -46,3 +46,12 @@ share_bones_time (Bone share time) int 1200 0
+
+#    If enabled, the reason why bones where dropped or not dropped, along
+#    with death location, is written to the minetest log.
+enable_bones_logging (Log location of deaths and reason if no bones) bool true
```
Suggestion: `Log bones actions`

poikilos replied:
sounds good

SmallJoker commented:
The code itself looks good now. I will approve this PR as soon these few comments are handled.

poikilos commented:
What do the other maintainers think of this?

Ezhh commented on mods/bones/init.lua:
```diff
@@ -186,20 +186,42 @@ minetest.register_on_dieplayer(function(player)
+		if enable_bones_logging then
+			minetest.log("action", "[bones] " .. player_name ..
+				"'s bones do not remain since player has creative -- " ..
+				"died at " .. minetest.pos_to_string(pos))
```
I find "bones do not remain" very awkward. It's also better to use "because" instead of "since". 
Perhaps: "player_name died at pos, but bones do not appear because player_name has creative."
This is also fixes the awkward way pos is tagged on the end right now.

These points (especially use of "since") apply to the other messages as well.

paramat replied:
I agree, 'bones do not remain' is bad English, and 'because' is better than 'since'.
Maybe 'bones are not placed' or something.

paramat replied:
Random suggestion:
(player) died at (pos). Bones were not placed because player has creative privilege

Ezhh commented on mods/bones/init.lua:
```diff
@@ -218,12 +240,33 @@ minetest.register_on_dieplayer(function(player)
+
+		if enable_bones_logging then
+			minetest.log("action", "[bones] " .. player_name ..
+				"'s bones do not remain since cannot place bones here -- " ..
```
Suggestion: "player_name died at pos, but bones do not appear because bones cannot be placed here."

Ezhh replied:
I really dislike "bones do not remain".  It comes across as poor english to me, instead of the way you intend it, and I don't see it adding to immersion. Maybe someone else can weigh in with their thoughts.

paramat replied:
'since / because cannot place bones here' is bad.

paramat replied:
`(player) died at (pos). Items are dropped because bones cannot be placed here`

paramat commented on mods/bones/init.lua:
```diff
@@ -218,12 +240,33 @@ minetest.register_on_dieplayer(function(player)
 
+	if enable_bones_logging then
+		minetest.log("action", "[bones] " .. player_name ..
+			"'s bones remain where died at " .. minetest.pos_to_string(pos))
```
'where died at' is bad too :]

paramat replied:
`(player) died at (pos) and bones were placed`

paramat commented:
Really needs better wording, maybe we can help with this.
I have no objection to the concept, it seems somewhat useful.

paramat commented:
Rethinking:
Since this is logging it needs to be concise, 1 line preferably. Also logging is present tense 'player places / digs ...', so a suggestion:

`(player) dies at (pos). No bones placed: creative`
`(player) dies at (pos). No bones placed: no inventory`
`(player) dies at (pos). No bones placed: placement disallowed, items dropped`
`(player) dies at (pos). Bones placed`

poikilos commented:
I commented on the previous suggestions, then saw paramat's suggestion which looks neat for logging.
So should I change to paramat's shorter messages or does anyone have a comment?

poikilos commented:
I fixed the conflict with settingstypes.txt. I'm sure I fixed that conflict correctly, but I'm not that familiar with rebasing--can anyone tell if that worked out ok?

paramat commented:
@Ezhh what do you think about my concise suggestion?

Ezhh commented:
It's good for the logs, but of course we need it in full for the in game messages.

paramat commented:
Yes the in-game versions could be more readable versions. Maybe stretch out the concise versions into more language-like versions.

poikilos commented:
For chat, either we could use the following phrases in parenthesis (or as a separate sentence if that is decided to be better):
<player> dies at <pos> (no bones were placed because of creative or keep mode).
<player> dies at <pos> (no bones were placed because no items were in inventory).
<player> dies at <pos> (inventory was dropped because of drop mode or non-placeable location).
<player> dies at <pos> (bones were placed).

Ezhh commented:
Can we please use what I suggested previously for the in game chat messages? It's great for logs to talk about modes, but not great at all to put that in in game chat messages for players.

poikilos commented:
I am ok with that. So, we are looking at the short messages above for log, but Ezhh's more flowing sentence structure for chat (I spelled out the additional message Ezhh didn't in case anyone has further comments):
1. player_name died at pos, but bones do not appear because player has creative, or keep mode is enabled.
2. player_name died at pos, but bones do not appear because no items were in inventory.
3. player_name died at pos, but inventory was dropped because bones cannot be placed here.
4. player_name died at pos, and bones were placed there.

## Split Version
However, I suggest to make the sentences less awkward I should just split off the if statements and instead of the first and last third above, do:
1. a. player_name died at pos, but bones do not appear because keep mode is enabled.
1. b. player_name died at pos, but bones do not appear because player has creative.
3. a. player_name died at pos, but inventory was dropped because drop mode is enabled.
3. b. player_name died at pos, but inventory was dropped because bones cannot be placed here.

(and after splitting the `if` statements, the log output could be also modified for each:)
```
1. a. <player> dies at <pos>. No bones placed: keep mode
1. b. <player> dies at <pos>. No bones placed: creative
3. a. <player> dies at <pos>. Inventory dropped: drop mode
3. b. <player> dies at <pos>. Inventory dropped: bones not placeable
```
The Split Version seems to make the output more readable in both log and chat messages. Since these messages would be something people always see, being clear seems helpful. I can easily make these changes. Are there any comments or should I go with the Split Version above? @Ezhh I changed "bones do not appear" to "inventory was dropped"  where applicable--let me know what you think.

Ezhh commented:
That still leaves references to modes in in game messages, which we shouldn't have. To simplify, how about the following?

Player drops bones:
player_name died at pos, and bones were placed there.

Player drops inventory:
player_name died at pos, and dropped their inventory.

Player dies but drops neither inventory nor bones (because they didn't have items):
player_name died at pos.

This is in game so doesn't need to be technical. It also doesn't need to explain why a certain thing happened. It just needs a simple short message to inform the player of what they should expect to find and where.

Keep the more technical details in the short form for the logs as suggested. Those are good.

paramat commented:
^ I agree with the previous post, good suggestions for the messages to the player.

poikilos commented:
That is a good solution. I'll start making changes soon.
Now we are looking at using the more specific ("Split Version") log entries, and the chat messages reduced to the 3 statements by Ezhh. Let me know if anyone else should weigh in on this--that would be helpful before I start, but I will also have time to make more changes later.

Ezhh commented:
You only need one more approval, and will probably get two more with these changes, so you should be safe to go ahead. :)

poikilos commented:
I implemented the following:

### Chat:
Player drops bones:
* player_name .. " died at " .. pos_string .. ", and bones were placed there."

Player drops inventory:
* player_name .. " died at " .. pos_string .. ", and dropped their inventory."

Player dies but drops neither inventory nor bones:
* player_name .. " died at " .. pos_string .. "."


### Log:
* player_name .. " dies at " .. pos_string .. ". No bones placed: keep mode"
* player_name .. " dies at " .. pos_string .. ". No bones placed: creative"
* player_name .. " dies at " .. pos_string .. ", but bones do not appear because no items were in inventory."
* player_name .. " dies at " .. pos_string .. ". Inventory dropped: drop mode"
* player_name .. " dies at " .. pos_string .. ". Inventory dropped: bones not placeable"
* player_name .. " dies at " .. pos_string .. ", and bones were placed there."

Do you think we should change the tense of the chat, or change the last log message to `player_name .. " dies at " .. pos_string .. ", and bones have been placed there."`, or do you think it is good?

SmallJoker commented:
@paramat The exact wording is bikeshedding. I'm not sure how precise you want to log a death, but I'm fine with both versions - old and new as long it works and the code is still formatted correctly.

poikilos commented:
I'm satisfied with the wording. I think I like the past tense the way it is for the chat since it makes more sense conversationally. I also like the present tense for the log since block interactions are logged that way. If anyone wants to take away the needs improvement tag or approve and merge this, feel free.

Ezhh commented:
Can't test at the moment, but updated messages look good.

paramat commented:
Chat messages are fine, and past tense is best for them, but 2 of the log messages are now too long, conversational and inconsistent,:
```
player_name .. " dies at " .. pos_string .. ", but bones do not appear because no items were in inventory."
player_name .. " dies at " .. pos_string .. ", and bones were placed there."
```
 i suggest:
```
    player_name .. " dies at " .. pos_string .. ". No bones placed: keep mode"
    player_name .. " dies at " .. pos_string .. ". No bones placed: creative"
    player_name .. " dies at " .. pos_string .. ", No bones placed: no inventory"
    player_name .. " dies at " .. pos_string .. ". Inventory dropped: drop mode"
    player_name .. " dies at " .. pos_string .. ". Inventory dropped: bones not placeable"
    player_name .. " dies at " .. pos_string .. ". Bones placed"
```
Logs need to be short to preferably fit on 1 line in a small terminal.

About number of different log messages: maybe there are too many? 'keep mode' and 'creative' seem redundant information because the server admin has already set those. 'no inventory' seems useless information to log.
Is distinguishing 'drop mode' from 'bones not placeable' useful information?
Input from server admin is appreciated on this.

So we perhaps only need 3:
```
(player) dies at (pos). No bones placed
(player) dies at (pos). Inventory dropped
(player) dies at (pos). Bones placed
```

tenplus1 commented:
@paramat - yup, keep it short and simple for the logs, the 3 simple examples are the way to go.

paramat commented:
Thanks, logging messages are fine for me. Let's see what Ezhh thinks.

Ezhh commented:
No issues for me. Making it cleaner/simpler is good.

paramat commented:
Assuming your approval then:
> This will have my support if these are adjusted.

paramat commented:
Just one thing: i don't see a need for the `enable_bones_logging` setting, if it's useful to log then it should always be logged. No other logging has individual options, the option is already in the 'log level' setting to choose how verbose logging is.

poikilos commented:
So does everyone agree on that? This is moved back to "One approval" status pending agreement.

Ezhh commented:
@paramat I don't see why anyone would be unhappy with it being always on... Players dying is not such a frequent event that it will fill the logs with spam. If you have tested and are happy with this, it's probably time to merge...

paramat commented:
Seems good and mergeable, i haven't tested though, will +1 once i have (or someone else has).

poikilos commented:
I tested it already in the way I mentioned but if anyone can help, please try drowning yourself with or without items in the different modes (remember to set `enable_bones_chat_msg = true` to your minetest.conf to test that). Here is a bash script to install my patched version (makes backup as long as you don't already have a file called init.1st in your minetest_game/mods/bones directory).

### Install patch
```bash
MT_GAME_PATH=/usr/local/share/minetest/games/minetest_game
if [ ! -d "$MT_GAME_PATH" ]; then
  echo "ERROR: nothing done since missing $MT_GAME_PATH"
  exit 1
fi
cd $MT_GAME_PATH/mods/bones
if [ ! -f init.1st ]; then
  sudo mv init.lua init.1st
else
  rm init.lua
fi
wget -O init.lua https://github.com/poikilos/minetest_game/raw/master/mods/bones/init.lua
```

### Revert from patch to minetest's version
```bash
MT_GAME_PATH=/usr/local/share/minetest/games/minetest_game
if [ ! -d "$MT_GAME_PATH" ]; then
  echo "ERROR: nothing done since missing $MT_GAME_PATH"
  exit 1
fi
cd $MT_GAME_PATH/mods/bones
if [ ! -f init.1st ]; then
  sudo mv init.lua init.1st
else
  rm init.lua
fi
wget -O init.lua https://github.com/minetest/minetest_game/raw/master/mods/bones/init.lua
```

paramat commented:
Seems well tested so :+1: will merge.

paramat commented:
Sorry for another request, a minor issue, this needs documenting in minetest.conf.example here with the other bones stuff https://github.com/minetest/minetest_game/blob/145b7bfef33eb640b5960fd9a8699af9302fe1d7/minetest.conf.example#L25
In settingtypes.txt the setting should be grouped with the other bones settings.

paramat commented on settingtypes.txt:
```diff
@@ -50,3 +50,6 @@ enable_stairs_replace_abm (Replace old stairs) bool false
 #    suitable starting biome.
 engine_spawn (Use engine spawn search) bool false
+
+#    Inform player of condition and location of new bones.
```
Needs moving to just after 'share_bones_time' above.

poikilos replied:
good catch

paramat commented:
Thanks code seems good.
I'm considering a better setting name and would like dev input. 'enable' is redundant for bool settings and just makes them longer, 'chat' is somewhat redundant next to 'msg', and it may be good to make the setting name more descriptive.
Maybe 'bones_position_message'?

poikilos commented:
The variable is renamed. I didn't know which way the project was going with boolean naming since most are enable_*. I can understand avoiding abbreviations for international projects.

paramat commented:
Thanks, all fine for me, will test now.
@Ezhh are you happy with the new setting name? If so are you +1 for the PR?

paramat commented:
Tested and fine.
"player dies at pos and bones were placed there." the 'there' at the end sounds a little awkward and is unnecessary.

paramat commented:
I noticed you are not using branches for your PRs, it will help you a lot to do so, see https://help.github.com/categories/bootcamp/ and after 'fork a repo' move onto https://help.github.com/articles/creating-and-deleting-branches-within-your-repository/

paramat commented:
:+1: I'll merge in 30 mins and edit the 'there' out when i merge.

paramat commented:
This page says no conflicts but due to nearby recent changes in settingtypes.txt this needs rebasing. If you find rebasing difficult you are welcome to open a new PR for this.

SmallJoker commented:
@paramat 
```bash
wget https://github.com/minetest/minetest_game/pull/2082.diff && git apply 2082.diff
# Removing "there"
git add <affected paths>
git commit -m "Add optional bones messages for player and log" --author "<author info from .patch file>"
git push
```
Works for me.
@poikilos I cleaned up your `master` branch. Might need re-cloning to local to ensure that it uses the correct commit history.

paramat commented:
Yes, if your fork was only for this PR.
